packages/forms/src/Concerns/InteractsWithForms.php

```diff
@@ -229,6 +229,19 @@ public function getActiveFormsLocale(): ?string
 
     public function updatingInteractsWithForms(string $statePath): void
     {
+        // is it UploadField with 3 part state?
```

Comment:
in `updatingInteractsWithForms `the index would be `data.myfile.f0f5612c-a7fd-4333-a0c9-c1b2cc4a58da` so is saves something like this in `oldFormState` array:

`$this->oldFormState['data.myfile.f0f5612c-a7fd-4333-a0c9-c1b2cc4a58da'] = data_get($this, 'data.myfile.f0f5612c-a7fd-4333-a0c9-c1b2cc4a58da');`

which would be

`$this->oldFormState['data.myfile.f0f5612c-a7fd-4333-a0c9-c1b2cc4a58da'] = null;`

bcos of that index. we need to correct the index when it save date into $this->oldFormState array.

to become:
`$this->oldFormState['data.myfile'] = [SOME DATA];`

Reply:
Other fields can have 3 part state, for example repeaters, builders, any JSON nesting. Also file uploads that are mounted to a Livewire property. We just cannot rely on this.

Reply:
The solution should be implemented somewhere in the BaseFileUpload class, you probably need to override `getOldState()`. Be aware that if it's `$this->isMultiple()`, we need to return an array of uploads, not just the first.

Reply:
How about just `$statePath = (string) str($statePath)->before('.');`

Which will then always store the most state that you could possibly need from the form, even if we don't need all of it

Reply:
It would replace all the code you added here

Reply:
That will return just first part :) 
`data.myfile.f0f5612c-a7fd-4333-a0c9-c1b2cc4a58da` => `data`

maybe something like this you prefer
`        $statePath = (string) str($statePath)->explode('.')->slice(0,2)->join('.');`

Yes that is cleaner but in conclusion you approve that the best place to handle this issue is in `updatingInteractsWithForms` right?

and also plz help me to test this type of field you mentioned:

>  file uploads that are mounted to a Livewire property

how can create such field?

Reply:
Oh! I think you mean we always keep all data of form! I did not read it carefully!

You mean it will be always something like this:

$this->oldFormState[$statePath] = data_get($this, 'data');

Reply:
Good, I need to test those other 3 parts to see how they works in this context. but for those 2 part upload field you mentioned => if they are 2 parts then there should not be any problem and they have correct `statePath` so they will save old state correctly as usual. is not true? I will test that too, but plz tell me how I should create such field? =>

> > file uploads that are mounted to a Livewire property

but about `getOldState()` => I dont think that can help bcos root of issue is that in the first place old state is not correctly saved in the `$this->oldFormState`  bcos of that 3 part incorrect index. so we should hook on whenever we are saving into `$this->oldFormState` not when we are getting from it. 
agree?

Reply:
What do you mean it is not correctly saved?

Reply:
Yes

Reply:
Ok, accordingly we should change `getOldFormState` to use `data_get` instead of getting it directly:

```
    public function getOldFormState(string $statePath): mixed
    {
        return data_get($this->oldFormState, $statePath);
    }
```

also I tested this approach with a fileupload field inside a repeater and all of them are getting their old and new state correctly.

Reply:
@danharrin as we've discussed the new state path always will be `data`, would you recommend that we do it like this 
```
    public function updatingInteractsWithForms(string $statePath): void
    {
        $this->oldFormState['data'] = data_get($this, 'data');
    }
```
or like this?
```
    public function updatingInteractsWithForms(string $statePath): void
    {
        $statePath = (string) str($statePath)->before('.');

        $this->oldFormState[$statePath] = data_get($this, $statePath);
    }
```

first one would make `$statePath` as an unused parameter and second one looks like doing extra and useless job 😄

Reply:
Second one, forms are not always mounted to `data`, they can have many different state paths